lightning/src/chain/channelmonitor.rs

```diff
@@ -152,7 +153,12 @@ pub enum ChannelMonitorUpdateErr {
 /// corrupted.
 /// Contains a human-readable error message.
 #[derive(Debug)]
-pub struct MonitorUpdateError(pub &'static str);
+pub struct MonitorUpdateError {
+	/// The human-readable error message string.
+	pub msg: &'static str,
+	/// A monitor update may have an error but still need to be persisted to disk.
+	pub persist_updated_monitor: bool
+}
```

Comment:
Lmk what you think of the naming.

Reply:
Also, w/r/t the naming of `ChannelMonitorUpdateErr` and `MonitorUpdateError` -- what do you think of just renaming `MonitorUpdateError` to `InternalMonitorErr`?

Reply:
Based on our offline discussion, I'm starting to become of the mind that these should be reduced to a single error type. My reasoning is as follows:
- `ChannelMonitor`'s methods can return errors.
- `chain::Watch`'s interface is defined in terms of `ChannelMonitor`, `ChannelMonitorUpdate`, and `ChannelMonitorUpdateErr`.
- `ChainMonitor` is an implementation of `chain::Watch`, but there may be other implementations of the trait.

Therefore, whether a `ChannelMonitor` must be persisted isn't necessarily always an internal detail handled by `ChainMonitor`; another implementation of `chain::Watch` or someone using `ChannelMonitor` directly would also need to do so.

My thought then would be to have one error type where the `PermanentFailure` variant has a bool indicating if it should be persisted.

That said, I'd like to get a better idea about where in the code that the "failure but persist" error will occur. I couldn't quite make that out from https://github.com/rust-bitcoin/rust-lightning/pull/681#discussion_r485875141. Additionally, the docs for `PermanentFailure` indicate it can be used in two different ways. This makes me wonder if it should then be two different variants (and if so would this would remove the need for the bool?).

Thoughts? Let me know if I misunderstood anything.

Reply:
Hmm, I think they're definitely very different in their purpose, though I don't necessarily think that implies they can't be the same type - 
 * `MonitorUpdateError` is an error type returned by RL to the user to indicate that the update provided to the monitor was somehow bogus. It instructs the user whether they should still persist the updated object to disk, but in any case the update was not fully applied and ready to move forward.
 * `ChannelMonitorUpdateErr` is an error type generated by the user instructing the `Channel{,Manager}` what the current state of the monitor update is. It is maybe more aptly named `ChannelMonitorUpdateStatus` (and replacing the `Result` with just a general done/in-flight/failed enum).

The should-persist flag is definitely very awkward, and we should take pains in the documentation to point out that it is *only* relevant if you are running in a multi-primary redundant `ChannelMonitor` setup where each `ChannelMonitor` can unilaterally broadcast a transaction. In other words, for most folks, we can pretend its not there. See #679 for the *one* case where its ever set/matters. For normal folks, the `MonitorUpdateError` is just a human-readable error string that's arguably only useful for opening an issue on our github. 

Indeed, for more complicated setups, arguably, they start to blend a bit more, but I do think there's value in them being different.

Reply:
That's a good observation. I like where you're going with making `ChannelMonitorUpdateErr` a status instead of an error.

May need to think on this more, but just to throw an idea out there: would it make sense to have the `chain::Watch` interface return something like `Result<ChannelMonitorUpdateStatus, MonitorUpdateError>`? Possibly renaming these in some way, of course.

Reply:
> `Result<ChannelMonitorUpdateStatus, MonitorUpdateError>`

Hmm, interesting point. I suppose we could, though it doesn't mean anything to the `ChannelManager` which receives it - `Err(_)` would be converted to a `PermanentFailure` status either way.

Reply:
> > `Result<ChannelMonitorUpdateStatus, MonitorUpdateError>`
> 
> Hmm, interesting point. I suppose we could, though it doesn't mean anything to the `ChannelManager` which receives it - `Err(_)` would be converted to a `PermanentFailure` status either way.

Not 100% sure if I follow that last point. Presumably, `ChannelManager` would process either the status if the `Result` is `ok` or the error if it is `err` using `handle_monitor_err!`.

It's the `ChainMonitor` that currently converts any `MonitorUpdateError` to a `PermanentFailure`. In what I'm floating, `ChainMonitor` would examine the `Result` and persist if needed and then simply return the `Result` as is back to `ChannelManager`.

Reply:
Right, my point was only that that would be communicating extra information to the `ChannelManager` that it doesnt care about - `PermanentFailure` is the same to it as `Err(_)`.